helper_thread.c

```diff
@@ -188,14 +193,36 @@ static uint8_t wait_for_action(int fd, unsigned int timeout_ms)
 	FD_SET(fd, &rfds);
 	FD_ZERO(&efds);
 	FD_SET(fd, &efds);
-	res = select(fd + 1, &rfds, NULL, &efds, &timeout);
+#ifdef CONFIG_HAVE_TIMERFD_CREATE
+	{
+		/*
+		 * If the timer frequency is 100 Hz, select() will round up
+		 * `timeout` to the next multiple of 1 / 100 Hz = 10 ms. Hence
+		 * use a high-resolution timer if possible to increase
+		 * select() timeout accuracy.
+		 */
+		struct itimerspec delta = {};
+
+		delta.it_value.tv_sec = timeout.tv_sec;
+		delta.it_value.tv_nsec = timeout.tv_usec * 1000;
+		res = timerfd_settime(timerfd, 0, &delta, NULL);
+		assert(res == 0);
+		FD_SET(timerfd, &rfds);
+	}
+#endif
```

Comment:
I feel a little sad seeing this in this particular file but there's absolutely nothing portable that provides improved granularity AND works on all platforms short of reaching for some giant library like libevent (and arguably non-Linux platforms are no worse off than today) without open coding. My only thought is: should "a more granular select" be pushed into platform specific code?

Reply:
Making select() more granular would be suboptimal because a generic implementation either would have to select a timerfd from a pool or dynamically create the timerfd. The code in this pull request only creates one timerfd per helper thread. I think a similar comment applies to poll() and epoll().

Reply:
Oof that would be very complicated... OK understood.